lib/auth/permissions.go

```diff
@@ -139,18 +139,25 @@ func (a *authorizer) authorizeRemoteUser(u RemoteUser) (*Context, error) {
 	if len(roleNames) == 0 {
 		return nil, trace.AccessDenied("no roles mapped for remote user %q from cluster %q with remote roles %v", u.Username, u.ClusterName, u.RemoteRoles)
 	}
-	// Set "logins" trait and "kubernetes_groups" for the remote user. This allows Teleport to work by
-	// passing exact logins, kubernetes groups and users to the remote cluster. Note that claims (OIDC/SAML)
-	// are not passed, but rather the exact logins, this is done to prevent
-	// leaking too much of identity to the remote cluster, and instead of focus
-	// on main cluster's interpretation of this identity
+	// Set internal traits for the remote user. This allows Teleport to work by
+	// passing exact logins, kubernetes users/groups and database users/names
+	// to the remote cluster.
 	traits := map[string][]string{
 		teleport.TraitLogins:     u.Principals,
 		teleport.TraitKubeGroups: u.KubernetesGroups,
 		teleport.TraitKubeUsers:  u.KubernetesUsers,
 		teleport.TraitDBNames:    u.DatabaseNames,
 		teleport.TraitDBUsers:    u.DatabaseUsers,
 	}
+	// Prior to Teleport 6.2 no user traits were passed to remote clusters
+	// except for those specified above. To preserve backwards compatible
+	// behavior, filter out any "internal" traits when applying traits from
+	// user identity.
+	for k, v := range u.Identity.Traits {
+		if !teleport.IsInternalTrait(k) {
```

Comment:
optional: you could check whether `k` is already in `traits` instead of using a helper function
that way you have 1 fewer places to maintain "internal" traits